src/test/groovy/graphql/DataLoaderCFTest.groovy

```diff
@@ -0,0 +1,287 @@
+package graphql
+
+import graphql.execution.instrumentation.dataloader.DataLoaderCF
+import graphql.schema.DataFetcher
+import org.dataloader.BatchLoader
+import org.dataloader.DataLoader
+import org.dataloader.DataLoaderFactory
+import org.dataloader.DataLoaderRegistry
+import spock.lang.Specification
+
+import java.util.concurrent.CompletableFuture
+
+import static graphql.ExecutionInput.newExecutionInput
+
+class DataLoaderCFTest extends Specification {
+
+
+    def "chained data loaders"() {
+        given:
+        def sdl = '''
+
+        type Query {
+          dogName: String
+          catName: String
+        }
+        '''
+        int batchLoadCalls = 0
+        BatchLoader<String, String> batchLoader = { keys ->
+            return CompletableFuture.supplyAsync {
+                batchLoadCalls++
+                Thread.sleep(250)
+                println "BatchLoader called with keys: $keys"
+                assert keys.size() == 2
+                return ["Luna", "Tiger"]
+            }
+        }
+
+        DataLoader<String, String> nameDataLoader = DataLoaderFactory.newDataLoader(batchLoader);
+
+        DataLoaderRegistry dataLoaderRegistry = new DataLoaderRegistry();
+        dataLoaderRegistry.register("name", nameDataLoader);
+
+        def df1 = { env ->
+            return DataLoaderCF.newDataLoaderCF(env, "name", "Key1").thenCompose {
+                result ->
+                    {
+                        return DataLoaderCF.newDataLoaderCF(env, "name", result)
```

Comment:
Rather than static `DataLoaderCF.newDataLoaderCF(env, "name", "Key1")` we should provide a "factory" helper in DataFetchingEnvironment that creates these chains and hence doesnt need env passed in

```
def df1 = { env ->
            return env.chainedDataLoaders().newDataLoaderCF("name", "Key1").thenCompose {
```

Reply:
Ok ... happy with that.